### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by themanforfree**
To avoid conflicts with KvServer's commands, AuthServer's commands has empty keys field and will always propose with slow path

### CODE REVIEW

File: xline/src/storage/authstore.rs [hunk 1]
```diff
@@ -0,0 +1,1459 @@
...
+            return false;
+        }
+        let user = User {
+            name: req.name.into_bytes(),
+            password: req.hashed_password.into_bytes(),
+            options: req.options,
+            roles: Vec::new(),
+        };
+        self.put_user(&user, revision, 0);
+        self.refresh_permission_cache();
```
Comment (rogercloud) [inline xline/src/storage/authstore.rs:1166]: Do we really need to refresh cache every time? Is there any cheaper way?

File: xline/src/storage/authstore/backend.rs [hunk 2]
```diff
@@ -0,0 +1,1235 @@
...
+                        "Failed to decode role from kv value, error: {:?}, kv: {:?}",
+                        e, kv
+                    )
+                })
+            })
+            .collect()
+    }
+
+    /// speculative execute command
+    pub(crate) fn speculative_exec(&self, execution_req: ExecutionRequest) {
```
Comment (rogercloud) [inline xline/src/storage/authstore/backend.rs:339]: Why do we call it "speculative_exec"? How is it speculative?

Comment (themanforfree) [reply xline/src/storage/authstore/backend.rs:339]: just use old code, command execution will be refactor in the future

File: xline/src/server/auth_server.rs [hunk 3]
```diff
@@ -0,0 +1,512 @@
...
+        Ok(tonic::Response::new(res))
+    }
+
+    async fn auth_status(
+        &self,
+        request: tonic::Request<AuthStatusRequest>,
+    ) -> Result<tonic::Response<AuthStatusResponse>, tonic::Status> {
+        debug!("Receive AuthStatusRequest {:?}", request);
+
+        let is_fast_path = true;
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/server/auth_server.rs:176]: use will choose use fast or slow path in the future, here is just a temporary implementation

File: xline/src/rpc/mod.rs [hunk 4]
```diff
@@ -71,40 +71,218 @@ mod leasepb {
...
     lock_server::{Lock, LockServer},
     LockRequest, LockResponse, UnlockRequest, UnlockResponse,
 };
 
+pub use self::etcdserverpb::range_request::{SortOrder, SortTarget};
+
+impl User {
+    /// Check if user has the given role
+    pub(crate) fn has_role(&self, role: &str) -> bool {
+        self.roles.binary_search(&role.to_owned()).is_ok()
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/rpc/mod.rs:112]: yes, we sort this vec when push new items

File: xline/src/server/auth_server.rs [hunk 5]
```diff
@@ -0,0 +1,512 @@
...
+        let (res, sync_res) = self.propose_slow_path(request).await?;
+
+        let mut res: AuthRoleAddResponse = res.decode().into();
+        let revision = sync_res.revision();
+        debug!("Get revision {:?} for AuthRoleAddResponse", revision);
+        if let Some(mut header) = res.header.as_mut() {
+            header.revision = revision;
+        }
+        Ok(tonic::Response::new(res))
+    }
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/server/auth_server.rs:399]: these functions are required by proto, and they just send propose to curp and wait for response

File: xline/src/storage/authstore/store.rs [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+            AuthRoleRevokePermissionRequest, AuthUserAddRequest, AuthUserDeleteRequest,
+            AuthUserGrantRoleRequest, Permission,
+        },
+        storage::authstore::perms::{PermissionCache, UserPermissions},
+    };
+
+    use super::*;
+
+    #[tokio::test]
+    async fn test_permisiion_cache() -> Result<(), Box<dyn Error>> {
```
Comment (rogercloud) [inline xline/src/storage/authstore/store.rs:303]: typo

File: xline/src/storage/authstore/store.rs [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+        for test in testcases() {
+            let (_, _) = send_req(&store, test.0).await?;
+            assert_eq!(store.inner.permission_cache(), test.1);
+        }
+
+        Ok(())
+    }
+
+    #[allow(clippy::too_many_lines)]
+    fn testcases() -> Vec<(RequestWithToken, PermissionCache)> {
```
Comment (rogercloud) [inline xline/src/storage/authstore/store.rs:317]: please separate the cases into different tests.

File: xline/src/utils.rs [hunk 8]
```diff
@@ -0,0 +1,45 @@
+use parking_lot::{RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+/// Apply a closure on a rwlock after getting the guard
+pub(crate) trait RwLockMap<T, R> {
```
Comment (rogercloud) [inline xline/src/utils.rs:4]: I think there's an util for the lock in CURP. shall we merge them?

File: xline/src/server/command.rs [hunk 9]
```diff
@@ -100,6 +108,43 @@ impl KeyRange {
...
+                (Bound::Unbounded, _) => true,
+                (_, Bound::Unbounded) => false,
+            };
+            s1_lt_s2 && e1_gt_e2
+        }
+    }
+
+    /// Get end of range with prefix
+    #[allow(clippy::indexing_slicing)] // end[i] is always valid
+    pub(crate) fn get_prefix(key: &[u8]) -> Vec<u8> {
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/server/command.rs:136]: Added comments

File: xline/src/rpc/mod.rs [hunk 10]
```diff
@@ -71,40 +71,218 @@ mod leasepb {
...
+    /// Kv backend
+    Kv,
+    /// Auth backend
+    Auth,
+}
+
+impl RequestWrapper {
+    /// Get the backend of the request
+    pub(crate) fn backend(&self) -> RequestBackend {
+        match *self {
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/rpc/mod.rs:237]: [clippy lints](https://rust-lang.github.io/rust-clippy/master/index.html#match_ref_pats)

File: xline/src/utils.rs (hunk 8 above)
Comment (themanforfree) [reply xline/src/utils.rs:4]: I will merge them to a new crate in next PR

File: xline/src/server/command.rs [hunk 11]
```diff
@@ -57,6 +57,14 @@ pub(crate) struct KeyRange {
 }
 
 impl KeyRange {
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/server/command.rs:59]: `KeyRange` will refactor when use new [merged_range](https://github.com/datenlord/merged_range) crate

File: xline/src/server/auth_server.rs [hunk 12]
```diff
@@ -0,0 +1,398 @@
...
+
+    /// Generate propose id
+    fn generate_propose_id(&self) -> ProposeId {
+        ProposeId::new(format!("{}-{}", self.name, Uuid::new_v4()))
+    }
+
+    /// Generate `Command` proposal from `Request`
+    fn command_from_request_wrapper(propose_id: ProposeId, wrapper: RequestWithToken) -> Command {
+        Command::new(vec![], wrapper, propose_id)
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/server/auth_server.rs:69]: Here is just move without heap allocation

**Review by rogercloud — APPROVED** [review]

**Pull request merged**